pkg/promclient/promclient.go

```diff
@@ -88,6 +89,159 @@ func ExternalLabels(ctx context.Context, logger log.Logger, base *url.URL) (labe
+		WebEnableAdminAPI  modelBool     `json:"web.enable-admin-api"`
+		WebEnableLifecycle modelBool     `json:"web.enable-lifecycle"`
+	}{}
+
+	if err := json.Unmarshal(b, &parsableFlags); err != nil {
+		return err
+	}
+
+	*f = Flags{
+		TSDBPath:           parsableFlags.TSDBPath,
+		TSDBRetention:      model.Duration(parsableFlags.TSDBRetention),
+		TSDBMinTime:        model.Duration(parsableFlags.TSDBMinTime),
+		TSDBMaxTime:        model.Duration(parsableFlags.TSDBMaxTime),
+		WebEnableAdminAPI:  bool(parsableFlags.WebEnableAdminAPI),
+		WebEnableLifecycle: bool(parsableFlags.WebEnableLifecycle),
+	}
+	return nil
+}
+
+type modelDuration model.Duration
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (d *modelDuration) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	dur, err := model.ParseDuration(s)
+	if err != nil {
+		return err
+	}
+	*d = modelDuration(dur)
+	return nil
+}
+
+type modelBool bool
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (m *modelBool) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	boolean, err := strconv.ParseBool(s)
+	if err != nil {
+		return err
+	}
+	*m = modelBool(boolean)
+	return nil
+}
+
+// ConfiguredFlags some configured flags from /api/v1/status/flags Prometheus endpoint.
+// Added to Prometheus from v2.2.
+func ConfiguredFlags(ctx context.Context, logger log.Logger, base *url.URL) (Flags, error) {
+	u := *base
+	u.Path = path.Join(u.Path, "/api/v1/status/flags")
+
+	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+	if err != nil {
+		return Flags{}, errors.Wrap(err, "create request")
+	}
+	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return Flags{}, errors.Wrapf(err, "request config against %s", u.String())
+	}
+	defer runutil.CloseWithLogOnErr(logger, resp.Body, "query body")
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return Flags{}, errors.Errorf("failed to read body")
+	}
+
+	if resp.StatusCode != 200 {
+		return Flags{}, errors.Errorf("got non-200 response code: %v, response: %v", resp.StatusCode, string(b))
+	}
+
+	var d struct {
+		Data Flags `json:"data"`
+	}
+	if err := json.Unmarshal(b, &d); err != nil {
+		return Flags{}, errors.Wrapf(err, "unmarshal response: %v", string(b))
+	}
+
+	return d.Data, nil
+}
+
+// Snapshot will request Prometheus to perform snapshot in directory returned by this function.
+// Returned directory is relative to Prometheus data-dir.
+// NOTE: `--web.enable-admin-api` flag has to be set on Prometheus.
+// Added to Prometheus from v2.1.
+// TODO(bwplotka): Add metrics.
+func Snapshot(ctx context.Context, logger log.Logger, base *url.URL, skipHead bool) (string, error) {
+	u := *base
+	u.Path = path.Join(u.Path, "/api/v1/admin/tsdb/snapshot")
+
+	req, err := http.NewRequest(
```

Comment:
might be worth to check the `ConfiguredFlags` for the web.enable-admin-api? 
If called more times it would cause overhead but on the other hand the check will provide always relevant hint what is the issue.
Or maybe just checking for 404 status on the napshot API url would be enough to give that hint.

Reply:
sooo we need enable admin api anyway (: There is no regression on that side.

Reply:
hm do you mean to use Configured Flags to check admin api flag e.g to tell if we can use snapshotting or not? 

That is the plan, just not on this PR (:

Reply:
Yep, that's exactly what I had in mind. Ok :)